<h1 style="text-align:center; 
           background:#d9ecf2; 
           color:#045174; 
           font-size:45px; 
           padding:20px;
           border-radius:12px;">
     <b> 🌧️ Rainfall & Climate Data Analysis</b>
</h1>

<p align="center">
  <img src="img\starting.gif" width="280">
</p>

### <b>Dataset – Overview</b>

This dataset contains historical climate, weather, and rainfall-related measurements recorded across different locations and ***years(1990-2025)***. 

The data is structured on a daily basis, where each row represents one day of recorded environmental conditions.

The dataset includes ***130,230 records*** and ***21 columns*** covering a wide range of atmospheric and rainfall-related variables. 

These features are commonly used in weather analysis, climate research, and predictive modeling such as rainfall forecasting.

### <b>🧑‍💼 Project Author</b>

***Author: Muhammad Usman***

***Data Science/Machine Learning/DataEngineering Enthusiast***

***Date: December 2025***

<h2 style="text-align:center; 
           background:#d9ecf2; 
           color:#045174; 
           font-size:45px; 
           padding:15px;
           border-radius:12px;
           display: inline-block">
     <b> Importing Libraries</b>
</h2>


In [1]:
# Data Manupilation
import pandas as pd
import numpy as np

# Data visulization
import seaborn as sns
import matplotlib.pyplot as plt
import plotly.express as px

%matplotlib inline
import warnings
warnings.filterwarnings("ignore")

<h2 style=" 
           background:#d9ecf2; 
           color:#045174; 
           font-size:45px; 
           padding:15px;
           border-radius:12px;
           display: inline-block;">
     <b>Data Loading | Exploring  | Cleaning</b>
</h2>


In [2]:
raw_data = pd.read_csv("lac data.csv")
raw_data

Locations  YEAR  DOY  ALLSKY_SFC_UV_INDEX  ALLSKY_SFC_SW_DNI  \
0             Gharo  1990    1               -999.0             -999.0   
1             Gharo  1990    2               -999.0             -999.0   
2             Gharo  1990    3               -999.0             -999.0   
3             Gharo  1990    4               -999.0             -999.0   
4             Gharo  1990    5               -999.0             -999.0   
...             ...   ...  ...                  ...                ...   
130225  mirpur khas  2025  235               -999.0             -999.0   
130226  mirpur khas  2025  236               -999.0             -999.0   
130227  mirpur khas  2025  237               -999.0             -999.0   
130228  mirpur khas  2025  238               -999.0             -999.0   
130229  mirpur khas  2025  239               -999.0             -999.0   

           T2M  T2MDEW  T2M_RANGE  T2M_MAX  T2M_MIN  ...    RH2M  PRECTOTCORR  \
0        18.73    4.88      14.20    26.01    11.81  ...   44.59         0.00   
1        20.27   13.82       9.82    26.12    16.30  ...   68.97         0.77   
2        20.79   15.31       8.84    26.23    17.39  ...   74.23         0.71   
3        20.75   11.12      12.41    27.60    15.19  ...   60.31         0.00   
4        19.84    7.14      13.23    27.31    14.08  ...   46.46         4.68   
...        ...     ...        ...      ...      ...  ...     ...          ...   
130225   29.24   24.90       4.86    32.35    27.49  ...   77.56         2.13   
130226   29.20   24.69       5.06    32.31    27.25  ...   76.80         1.30   
130227   29.14   24.54       5.42    32.58    27.16  ...   76.46         0.79   
130228 -999.00 -999.00    -999.00  -999.00  -999.00  ... -999.00      -999.00   
130229 -999.00 -999.00    -999.00  -999.00  -999.00  ... -999.00      -999.00   

        IMERG_PRECTOT      PS  WS2M_MAX  WS2M_MIN  WS2M_RANGE   WD2M  \
0                -999  101.49      2.85      1.75        1.10  340.6   
1                -999  101.40      4.45      0.84        3.61  264.8   
2                -999  101.44      4.07      1.18        2.89  263.3   
3                -999  101.65      3.46      0.86        2.60    9.9   
4                -999  101.76      5.16      1.24        3.92   46.1   
...               ...     ...       ...       ...         ...    ...   
130225           -999   99.79      6.12      4.53        1.59  245.7   
130226           -999   99.87      5.71      4.06        1.65  245.0   
130227           -999   99.91      4.69      2.95        1.74  241.0   
130228           -999 -999.00   -999.00   -999.00     -999.00 -999.0   
130229           -999 -999.00   -999.00   -999.00     -999.00 -999.0   

        WS10M_MIN  T2MWET  
0            2.08   11.81  
1            0.97   17.04  
2            1.73   18.05  
3            1.28   15.94  
4            2.32   13.49  
...           ...     ...  
130225       6.13   27.07  
130226       5.51   26.94  
130227       3.97   26.84  
130228    -999.00 -999.00  
130229    -999.00 -999.00  

[130230 rows x 21 columns]

In [3]:
# show max columns

pd.set_option('display.max_columns', None)

* <b> Make the copy of original dataset into a new variable, to keep both cleaned and raw dataset separately

In [4]:
cleaned_data = raw_data.copy()

### <b> Dataset Column Descriptions <b>

| **Column Name**         | **Description**                                                           |
| ----------------------- | ------------------------------------------------------------------------- |
| **Locations**           | Geographical location or weather station where the data was recorded.     |
| **YEAR**                | Year of the observation.                                                  |
| **DOY**                 | Day of the year (1–365/366).                                              |
| **ALLSKY_SFC_UV_INDEX** | UV index at the surface under all-sky conditions (including clouds).      |
| **ALLSKY_SFC_SW_DNI**   | Direct Normal Irradiance; solar energy received per unit area.            |
| **T2M**                 | Air temperature at 2 meters above ground (°C).                            |
| **T2MDEW**              | Dew point temperature at 2 meters (°C).                                   |
| **T2M_RANGE**           | Daily temperature range (T2M_MAX − T2M_MIN).                              |
| **T2M_MAX**             | Maximum temperature of the day (°C).                                      |
| **T2M_MIN**             | Minimum temperature of the day (°C).                                      |
| **QV2M**                | Specific humidity at 2 meters (g/kg).                                     |
| **RH2M**                | Relative humidity at 2 meters (%).                                        |
| **PRECTOTCORR**         | Corrected precipitation (rainfall) in millimeters.                        |
| **IMERG_PRECTOT**       | Satellite-estimated rainfall from NASA IMERG (mm); missing values = -999. |
| **PS**                  | Surface atmospheric pressure (kPa).                                       |
| **WS2M_MAX**            | Maximum wind speed at 2 meters height (m/s).                              |
| **WS2M_MIN**            | Minimum wind speed at 2 meters height (m/s).                              |
| **WS2M_RANGE**          | Range of wind speed at 2 meters (WS2M_MAX − WS2M_MIN).                    |
| **WD2M**                | Wind direction at 2 meters height (degrees).                              |
| **WS10M_MIN**           | Minimum wind speed at 10 meters height (m/s).                             |
| **T2MWET**              | Wet-bulb temperature (combined effect of temperature + humidity).         |


<h2 style="text-align:center; 
           background:#d9ecf2; 
           color:#045174; 
           font-size:45px; 
           padding:15px;
           border-radius:12px;
           display: inline-block">
     <b> Analyzing Raw Data</b>
</h2>


<p align="center">
  <img src="img\rawdata.png" width="290", height="200">
</p>

* <b>To view top rows by default it prints first 5 rows

In [5]:
raw_data.head()

Locations  YEAR  DOY  ALLSKY_SFC_UV_INDEX  ALLSKY_SFC_SW_DNI    T2M  T2MDEW  \
0     Gharo  1990    1               -999.0             -999.0  18.73    4.88   
1     Gharo  1990    2               -999.0             -999.0  20.27   13.82   
2     Gharo  1990    3               -999.0             -999.0  20.79   15.31   
3     Gharo  1990    4               -999.0             -999.0  20.75   11.12   
4     Gharo  1990    5               -999.0             -999.0  19.84    7.14   

   T2M_RANGE  T2M_MAX  T2M_MIN   QV2M   RH2M  PRECTOTCORR  IMERG_PRECTOT  \
0      14.20    26.01    11.81   5.86  44.59         0.00           -999   
1       9.82    26.12    16.30   9.80  68.97         0.77           -999   
2       8.84    26.23    17.39  10.82  74.23         0.71           -999   
3      12.41    27.60    15.19   8.49  60.31         0.00           -999   
4      13.23    27.31    14.08   6.34  46.46         4.68           -999   

       PS  WS2M_MAX  WS2M_MIN  WS2M_RANGE   WD2M  WS10M_MIN  T2MWET  
0  101.49      2.85      1.75        1.10  340.6       2.08   11.81  
1  101.40      4.45      0.84        3.61  264.8       0.97   17.04  
2  101.44      4.07      1.18        2.89  263.3       1.73   18.05  
3  101.65      3.46      0.86        2.60    9.9       1.28   15.94  
4  101.76      5.16      1.24        3.92   46.1       2.32   13.49

* <b>To view last rows by default it prints 5 last rows

In [6]:
raw_data.tail()

Locations  YEAR  DOY  ALLSKY_SFC_UV_INDEX  ALLSKY_SFC_SW_DNI  \
130225  mirpur khas  2025  235               -999.0             -999.0   
130226  mirpur khas  2025  236               -999.0             -999.0   
130227  mirpur khas  2025  237               -999.0             -999.0   
130228  mirpur khas  2025  238               -999.0             -999.0   
130229  mirpur khas  2025  239               -999.0             -999.0   

           T2M  T2MDEW  T2M_RANGE  T2M_MAX  T2M_MIN    QV2M    RH2M  \
130225   29.24   24.90       4.86    32.35    27.49   19.78   77.56   
130226   29.20   24.69       5.06    32.31    27.25   19.51   76.80   
130227   29.14   24.54       5.42    32.58    27.16   19.32   76.46   
130228 -999.00 -999.00    -999.00  -999.00  -999.00 -999.00 -999.00   
130229 -999.00 -999.00    -999.00  -999.00  -999.00 -999.00 -999.00   

        PRECTOTCORR  IMERG_PRECTOT      PS  WS2M_MAX  WS2M_MIN  WS2M_RANGE  \
130225         2.13           -999   99.79      6.12      4.53        1.59   
130226         1.30           -999   99.87      5.71      4.06        1.65   
130227         0.79           -999   99.91      4.69      2.95        1.74   
130228      -999.00           -999 -999.00   -999.00   -999.00     -999.00   
130229      -999.00           -999 -999.00   -999.00   -999.00     -999.00   

         WD2M  WS10M_MIN  T2MWET  
130225  245.7       6.13   27.07  
130226  245.0       5.51   26.94  
130227  241.0       3.97   26.84  
130228 -999.0    -999.00 -999.00  
130229 -999.0    -999.00 -999.00

* <b>To view random rows

In [7]:
raw_data.sample(4)

Locations  YEAR  DOY  ALLSKY_SFC_UV_INDEX  ALLSKY_SFC_SW_DNI    T2M  \
94188        Quetta  1998  106              -999.00            -999.00  20.24   
67753       Larkana  1997   82              -999.00            -999.00  23.85   
127535  mirpur khas  2018  102                 2.34              20.16  31.02   
2095          Gharo  1995  270              -999.00            -999.00  30.30   

        T2MDEW  T2M_RANGE  T2M_MAX  T2M_MIN   QV2M   RH2M  PRECTOTCORR  \
94188    -5.71      16.07    28.48    12.41   2.98  17.58         0.00   
67753     2.20      19.12    33.69    14.57   4.52  27.71         0.05   
127535   18.00      15.28    39.75    24.47  14.06  57.72         0.01   
2095     21.25      10.32    36.45    26.13  15.90  61.76         0.00   

        IMERG_PRECTOT      PS  WS2M_MAX  WS2M_MIN  WS2M_RANGE   WD2M  \
94188            -999   82.20      2.60      0.22        2.38  219.6   
67753            -999  100.41      2.72      0.75        1.97  239.5   
127535           -999  100.46      6.31      2.35        3.96  276.5   
2095             -999  100.61      4.64      1.62        3.02  246.1   

        WS10M_MIN  T2MWET  
94188        0.34    7.26  
67753        0.89   13.03  
127535       3.08   24.51  
2095         1.90   25.78

* <b>Size of dataframe

In [8]:
num_rows = raw_data.shape[0]
num_columns = raw_data.shape[1]
shape_df = pd.DataFrame({

    'Description': ['Number of Rows', 'Number of Columns'],
    'Count': [num_rows, num_columns]
    
})
shape_df

Description   Count
0     Number of Rows  130230
1  Number of Columns      21

* <b>To get concise summary of whole dataset, including datatypes

In [9]:
raw_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 130230 entries, 0 to 130229
Data columns (total 21 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   Locations            130230 non-null  object 
 1   YEAR                 130230 non-null  int64  
 2   DOY                  130230 non-null  int64  
 3   ALLSKY_SFC_UV_INDEX  130230 non-null  float64
 4   ALLSKY_SFC_SW_DNI    130230 non-null  float64
 5   T2M                  130230 non-null  float64
 6   T2MDEW               130230 non-null  float64
 7   T2M_RANGE            130230 non-null  float64
 8   T2M_MAX              130230 non-null  float64
 9   T2M_MIN              130230 non-null  float64
 10  QV2M                 130230 non-null  float64
 11  RH2M                 130230 non-null  float64
 12  PRECTOTCORR          130230 non-null  float64
 13  IMERG_PRECTOT        130230 non-null  int64  
 14  PS                   130230 non-null  float64
 15  WS2M_MAX         

* <b>To generate summary statistics like mean, min, max, and quartiles for numerical columns.</b>

In [10]:
raw_data.describe().T

count         mean         std     min      25%  \
YEAR                 130230.0  2007.331030   10.294140  1990.0  1998.00   
DOY                  130230.0   181.970283  105.228068     1.0    91.00   
ALLSKY_SFC_UV_INDEX  130230.0  -313.868851  464.999998  -999.0  -999.00   
ALLSKY_SFC_SW_DNI    130230.0  -303.394025  472.149569  -999.0  -999.00   
T2M                  130230.0    25.917356   14.420308  -999.0    22.01   
T2MDEW               130230.0    11.976090   16.544042  -999.0     3.85   
T2M_RANGE            130230.0    12.350449   13.227072  -999.0     9.25   
T2M_MAX              130230.0    32.796289   14.523265  -999.0    28.98   
T2M_MIN              130230.0    20.292420   14.549809  -999.0    15.94   
QV2M                 130230.0    10.869556   13.856209  -999.0     5.41   
RH2M                 130230.0    50.037745   24.543912  -999.0    32.96   
PRECTOTCORR          130230.0     0.567021   13.248656  -999.0     0.00   
IMERG_PRECTOT        130230.0  -999.000000    0.000000  -999.0  -999.00   
PS                   130230.0    97.815896   14.707192  -999.0    99.07   
WS2M_MAX             130230.0     4.451676   12.548863  -999.0     3.31   
WS2M_MIN             130230.0     1.590123   12.481432  -999.0     0.68   
WS2M_RANGE           130230.0     2.708132   12.464430  -999.0     2.07   
WD2M                 130230.0   207.994721   97.834107  -999.0   151.70   
WS10M_MIN            130230.0     2.294816   12.540602  -999.0     0.98   
T2MWET               130230.0    18.946722   14.982753  -999.0    13.47   

                         50%      75%      max  
YEAR                 2007.00  2016.00  2025.00  
DOY                   181.00   273.00   366.00  
ALLSKY_SFC_UV_INDEX     1.26     2.00     3.53  
ALLSKY_SFC_SW_DNI      12.60    20.02    44.46  
T2M                    27.24    30.52    44.12  
T2MDEW                 14.93    21.58    28.45  
T2M_RANGE              12.97    15.79    24.45  
T2M_MAX                33.03    37.27    53.01  
T2M_MIN                21.75    26.31    37.23  
QV2M                   11.11    16.50    24.73  
RH2M                   52.29    68.21    95.47  
PRECTOTCORR             0.00     0.00   466.31  
IMERG_PRECTOT        -999.00  -999.00  -999.00  
PS                    100.06   100.82   102.24  
WS2M_MAX                4.47     5.77    14.02  
WS2M_MIN                1.27     2.50     9.04  
WS2M_RANGE              2.70     3.47    12.19  
WD2M                  242.70   270.80   360.00  
WS10M_MIN               2.09     3.46    12.10  
T2MWET                 21.21    25.99    32.13

* <b>To identify missing values and their % in each columns

In [11]:
missing_val = raw_data.isnull().sum()
missing_val_per = raw_data.isnull().sum()/len(raw_data) * 100

missing_data = pd.DataFrame({
    'Missing Values': missing_val,
    '% of Percentage': missing_val_per
})

missing_data

Missing Values  % of Percentage
Locations                         0              0.0
YEAR                              0              0.0
DOY                               0              0.0
ALLSKY_SFC_UV_INDEX               0              0.0
ALLSKY_SFC_SW_DNI                 0              0.0
T2M                               0              0.0
T2MDEW                            0              0.0
T2M_RANGE                         0              0.0
T2M_MAX                           0              0.0
T2M_MIN                           0              0.0
QV2M                              0              0.0
RH2M                              0              0.0
PRECTOTCORR                       0              0.0
IMERG_PRECTOT                     0              0.0
PS                                0              0.0
WS2M_MAX                          0              0.0
WS2M_MIN                          0              0.0
WS2M_RANGE                        0              0.0
WD2M                              0              0.0
WS10M_MIN                         0              0.0
T2MWET                            0              0.0

Hurray ! There is no missing values in the dataset :))...

but wait a minute, there are some columns which containing -999 as values which is not possible in real world scenario.
So, let's replace -999 with NaN values for better analysis.

* <b>Analyze missing values through graphs

In [12]:
# Filter columns with missing values > 0
missing_plot = missing_data[(missing_data['Missing Values'] > 0) | (missing_data['% of Percentage'] > 0)]

fig = px.bar(
    missing_plot,
    x = missing_plot.index,
    y = 'Missing Values',
    text = missing_plot['% of Percentage'].apply(lambda x: f"{x:.2f}%"),
    title = 'Missing Data Visualization',
    labels={'x': 'Columns', 'y': 'Number of Missing Values'}
)


fig.show()


## <b>Observations </b>

---

1. This dataset contains 21 columns and 130230 entries.

2. From descriptive analysis we observe that the rainfall dataset contains 1990-2025 ie 35 years of data

3. The dataset contains both numerical and categorical features.

4. Location column having object datatype which needs to be converted to categorical datatype.

5. Columns naming convention is not uniform. 

6. Feature Transformation is required for some columns, handle inconsistant data some values are in negative i.e is not real values we have to convert it in NAN or replace it by positive corrected values.

7. DOY (date of year) should be converted into a proper date format.


<h2 style=" 
           background:#d9ecf2; 
           color:#045174; 
           font-size:45px; 
           padding:15px;
           border-radius:12px;
           display: inline-block;">
     <b>Transforming</b>
</h2>

<p align="center">
  <img src="img\t.png" width="290", height="200">
</p>


This is the 2nd step of ETL after extraction data from websit...

* ***Renaming Columns***

In [13]:
# All abbreviated column names were transformed into human-readable standardized names following industry-style data preprocessing practices.

cleaned_data =  cleaned_data.rename(columns={
    
    "Locations": "Location",
    "YEAR": "Year",
    "DOY": "Day_Of_Year",
    "ALLSKY_SFC_UV_INDEX": "Uv_Index",
    "ALLSKY_SFC_SW_DNI": "Solar_Radiation_Dni",
    "T2M": "Temperature_Avg",
    "T2MDEW": "Dew_Point_Temp",
    "T2M_RANGE": "Temperature_Range",
    "T2M_MAX": "Temperature_Max",
    "T2M_MIN": "Temperature_Min",
    "QV2M": "Specific_Humidity",
    "RH2M": "Relative_Humidity",
    "PRECTOTCORR": "Rainfall_Mm",
    "IMERG_PRECTOT": "Satellite_Rainfall_Mm",
    "PS": "Surface_Pressure",
    "WS2M_MAX": "Wind_Speed_Max_2m",
    "WS2M_MIN": "Wind_Speed_Min_2m",
    "WS2M_RANGE": "Wind_Speed_Range_2m",
    "WD2M": "Wind_Direction_Deg",
    "WS10M_MIN": "Wind_Speed_Min_10m",
    "T2MWET": "Wet_Bulb_Temperature"
})

cleaned_data.head()

Location  Year  Day_Of_Year  Uv_Index  Solar_Radiation_Dni  Temperature_Avg  \
0    Gharo  1990            1    -999.0               -999.0            18.73   
1    Gharo  1990            2    -999.0               -999.0            20.27   
2    Gharo  1990            3    -999.0               -999.0            20.79   
3    Gharo  1990            4    -999.0               -999.0            20.75   
4    Gharo  1990            5    -999.0               -999.0            19.84   

   Dew_Point_Temp  Temperature_Range  Temperature_Max  Temperature_Min  \
0            4.88              14.20            26.01            11.81   
1           13.82               9.82            26.12            16.30   
2           15.31               8.84            26.23            17.39   
3           11.12              12.41            27.60            15.19   
4            7.14              13.23            27.31            14.08   

   Specific_Humidity  Relative_Humidity  Rainfall_Mm  Satellite_Rainfall_Mm  \
0               5.86              44.59         0.00                   -999   
1               9.80              68.97         0.77                   -999   
2              10.82              74.23         0.71                   -999   
3               8.49              60.31         0.00                   -999   
4               6.34              46.46         4.68                   -999   

   Surface_Pressure  Wind_Speed_Max_2m  Wind_Speed_Min_2m  \
0            101.49               2.85               1.75   
1            101.40               4.45               0.84   
2            101.44               4.07               1.18   
3            101.65               3.46               0.86   
4            101.76               5.16               1.24   

   Wind_Speed_Range_2m  Wind_Direction_Deg  Wind_Speed_Min_10m  \
0                 1.10               340.6                2.08   
1                 3.61               264.8                0.97   
2                 2.89               263.3                1.73   
3                 2.60                 9.9                1.28   
4                 3.92                46.1                2.32   

   Wet_Bulb_Temperature  
0                 11.81  
1                 17.04  
2                 18.05  
3                 15.94  
4                 13.49

* ***Converting Dtype***

In [14]:
# converting data types of Location column to category

cleaned_data['Location'] = cleaned_data['Location'].astype('category')

<h3 style=" 
           background:#d9ecf2; 
           color:#045270; 
           font-size:40px; 
           padding:10px;
           border-radius:12px;
           display: inline-block;">
     <b>Date Time Series</b>
</h3>


In [15]:
# Creating date series column

cleaned_data["Date"] = pd.to_datetime(cleaned_data["Year"].astype(str) + "-" + cleaned_data["Day_Of_Year"].astype(str), 
                             format="%Y-%j")

# dont rerun this cell buz we have drop Day_Of_Year column

In [16]:
# Extracting Month with names
cleaned_data["Month"] = cleaned_data["Date"].dt.month_name().astype('category')

# Extracting Day with names
cleaned_data["Day"] = cleaned_data["Date"].dt.day_name().astype('category')

In [17]:
# Convert Month name → month number
cleaned_data["Month_Num"] = pd.to_datetime(
    cleaned_data["Month"], format="%B"
).dt.month

In [18]:
# Drop the column "Day_Of_Year" buz we have done with feature Engineering
cleaned_data = cleaned_data.drop(columns=["Day_Of_Year"])

# if u re-run this code it will give error because Day_Of_Year column is deleted

In [19]:
# cleaned_data = cleaned_data.drop(columns=["Date"])

<h3 style=" 
           background:#d9ecf2; 
           color:#045270; 
           font-size:40px; 
           padding:10px;
           border-radius:12px;
           display: inline-block;">
     <b>Data Validation: Analysis of Negative and Positive Values</b>
</h3>


<b>While analyzing the dataset, negative values were observed in certain features.</b>

<b>This section validates which negative values are physically meaningful and which indicate data quality issues.</b>

In [20]:
cleaned_data.sample(5)

Location  Year  Uv_Index  Solar_Radiation_Dni  Temperature_Avg  \
61350       Karachi  2015      2.29                15.65            28.77   
75035       Larkana  2017      1.52                17.88            24.90   
27800     Hyderabad  1994   -999.00              -999.00            23.84   
127192  mirpur khas  2017      2.54                17.70            30.82   
105607       Thatta  1993   -999.00              -999.00            22.87   

        Dew_Point_Temp  Temperature_Range  Temperature_Max  Temperature_Min  \
61350            24.06               5.19            31.85            26.66   
75035             4.05              16.59            32.81            16.22   
27800            21.29               6.45            27.50            21.05   
127192           21.35              11.52            37.54            26.02   
105607           -3.26              18.23            33.35            15.12   

        Specific_Humidity  Relative_Humidity  Rainfall_Mm  \
61350               18.91              77.01         0.17   
75035                5.12              27.42         0.59   
27800               16.63              86.13         7.49   
127192              16.24              62.72         0.00   
105607               2.94              19.17         0.00   

        Satellite_Rainfall_Mm  Surface_Pressure  Wind_Speed_Max_2m  \
61350                    -999            100.06               7.17   
75035                    -999            100.75               3.39   
27800                    -999             95.60               2.20   
127192                   -999            100.79               6.53   
105607                   -999            101.73               4.25   

        Wind_Speed_Min_2m  Wind_Speed_Range_2m  Wind_Direction_Deg  \
61350                2.94                 4.23               245.4   
75035                0.18                 3.21               107.3   
27800                0.13                 2.07                70.5   
127192               2.28                 4.25               268.9   
105607               1.32                 2.93                36.1   

        Wind_Speed_Min_10m  Wet_Bulb_Temperature       Date     Month  \
61350                 3.72                 26.42 2015-05-08       May   
75035                 0.22                 14.48 2017-02-28  February   
27800                 0.17                 22.57 1994-10-21   October   
127192                3.05                 26.08 2017-05-04       May   
105607                2.96                  9.80 1993-11-24  November   

              Day  Month_Num  
61350      Friday          5  
75035     Tuesday          2  
27800      Friday         10  
127192   Thursday          5  
105607  Wednesday         11

* **Columns That Can Contain Both Positive and Negative Values**

| **Column Name**            | **Reason**                                                    |
| ---------------------- | ---------------------------------------------------------- |
| `Temperature_Avg`      | Average air temperature can fall below 0°C                 |
| `Temperature_Max`      | Maximum daily temperature may be below 0°C in cold regions |
| `Temperature_Min`      | Minimum daily temperature often goes below 0°C             |
| `Dew_Point_Temp`       | Dew point temperature can be negative                      |
| `Wet_Bulb_Temperature` | Can be negative under cold and dry conditions              |


These variables represent temperature-related measurements. 

Negative values are valid and indicate cold climatic conditions.

* **Columns That Must Always Be Positive or Zero**

(Negative values are physically impossible)

| Column Name             | Valid Range                            |
| ----------------------- | -------------------------------------- |
| `Rainfall_Mm`           | ≥ 0 mm                                 |
| `Satellite_Rainfall_Mm` | ≥ 0 mm (`-999` indicates missing data) |
| `Relative_Humidity`     | 0–100 %                                |
| `Specific_Humidity`     | ≥ 0                                    |
| `Surface_Pressure`      | > 0                                    |
| `Wind_Speed_Max_2m`     | ≥ 0                                    |
| `Wind_Speed_Min_2m`     | ≥ 0                                    |
| `Wind_Speed_Range_2m`   | ≥ 0                                    |
| `Wind_Speed_Min_10m`    | ≥ 0                                    |
| `Uv_Index`              | ≥ 0                                    |
| `Solar_Radiation_Dni`   | ≥ 0                                    |


| Column Name(Ranges- derived columns)           | Reason     |
| --------------------- | ------------------------------------------- |
| `Temperature_Range`   | Calculated as Max − Min; cannot be negative |
| `Wind_Speed_Range_2m` | Difference between max and min wind speed   |


These variables represent humidity, rainfall, wind, radiation, ranges, and pressure, which cannot be negative.

* <b>Vizulizing -ve values through graph</b>

In [21]:
# Select only numeric columns
numeric_df = cleaned_data.select_dtypes(include=[np.number])

negative_counts = (numeric_df < 0).sum().reset_index()
negative_counts.columns = ["Column_Name", "Negative_Value_Count"]

# Keep only columns that actually have negative values
negative_counts = negative_counts[negative_counts["Negative_Value_Count"] > 0]


fig = px.bar(

    negative_counts,
    x="Column_Name",
    y="Negative_Value_Count",
    
    title="Columns Containing Negative Values in the Dataset",
    labels={
        "Column_Name": "Feature Name",
        "Negative_Value_Count": "Count of Negative Values"
    }
)

fig.show()


***Explanation***

The bar chart displays the `count of negative values` across `numerical features` in the dataset. 

Initially, some columns appeared missing because large differences in negative value counts caused smaller bars to be visually compressed. 

After verifying the counts individually, it was confirmed that multiple features contain negative values. 

To clearly display all affected columns, a `logarithmic (log-scale) visualization` was used, which is a standard practice for `handling values with large magnitude differences`.

In [22]:
negative_counts

Column_Name  Negative_Value_Count
1                Uv_Index                 41070
2     Solar_Radiation_Dni                 41070
3         Temperature_Avg                   263
4          Dew_Point_Temp                 22400
5       Temperature_Range                    20
6         Temperature_Max                    29
7         Temperature_Min                  1883
8       Specific_Humidity                    20
9       Relative_Humidity                    20
10            Rainfall_Mm                    20
11  Satellite_Rainfall_Mm                130230
12       Surface_Pressure                    20
13      Wind_Speed_Max_2m                    20
14      Wind_Speed_Min_2m                    20
15    Wind_Speed_Range_2m                    20
16     Wind_Direction_Deg                    20
17     Wind_Speed_Min_10m                    20
18   Wet_Bulb_Temperature                  2732

* **Columns Containing Negative Values (Log Scale)**

In [23]:
fig = px.bar(
    negative_counts,
    x="Column_Name",
    y="Negative_Value_Count",
    title="Columns Containing Negative Values (Log Scale)",
    labels={
        "Column_Name": "Feature Name",
        "Negative_Value_Count": "Count of Negative Values"
    },
    log_y=True
)

fig.update_layout(
    height=700,
    xaxis_tickangle=-45
)

fig.show()


In [24]:
cleaned_data.sample(4)

Location  Year  Uv_Index  Solar_Radiation_Dni  \
104002                  Quetta  2025      0.52                 0.01   
80589                Nawabshah  1996   -999.00              -999.00   
13612   Hub LasbelaBalochistan  1991   -999.00              -999.00   
110978                  Thatta  2008      2.14                 9.87   

        Temperature_Avg  Dew_Point_Temp  Temperature_Range  Temperature_Max  \
104002             7.48            4.54               6.83            11.09   
80589             33.86           16.84              17.29            43.22   
13612             27.84           22.57               4.40            30.56   
110978            29.45           23.86               8.29            34.58   

        Temperature_Min  Specific_Humidity  Relative_Humidity  Rainfall_Mm  \
104002             4.26               6.48              82.52        16.28   
80589             25.93              12.28              41.82         0.00   
13612             26.16              17.29              73.58         0.00   
110978            26.29              18.65              73.24         0.02   

        Satellite_Rainfall_Mm  Surface_Pressure  Wind_Speed_Max_2m  \
104002                   -999             81.94               5.66   
80589                    -999             99.56               6.31   
13612                    -999             99.49               6.63   
110978                   -999             99.70               7.17   

        Wind_Speed_Min_2m  Wind_Speed_Range_2m  Wind_Direction_Deg  \
104002               1.02                 4.64               258.1   
80589                3.01                 3.30               213.8   
13612                4.41                 2.22               245.6   
110978               4.39                 2.78               250.6   

        Wind_Speed_Min_10m  Wet_Bulb_Temperature       Date      Month  \
104002                2.04                  6.01 2025-02-27   February   
80589                 4.22                 25.35 1996-09-17  September   
13612                 5.42                 25.20 1991-08-13     August   
110978                6.05                 26.65 2008-08-08     August   

             Day  Month_Num  
104002  Thursday          2  
80589    Tuesday          9  
13612    Tuesday          8  
110978    Friday          8

***Through graph we analyze column `Satelliterainfall_Mm` containing negative values that are invalid, so we can drop immediately...***

`Uv_index`, `Solar_Radiation_Dni`, `Specific_Humidity`, `Relative_Humidity`, `Rainfall_Mm`, `Surface_Pressure`, `Wind_Speed_Max_2m`, `Wind_Speed_Min_2m`, `Wind_Speed_Range_2m`, `Wind_Direction_Deg`, `Wind_Speed_Min_10m`

has some negative values, we will replace it Nan, then handle NAN values..

In [25]:
cleaned_data.drop(columns=["Satellite_Rainfall_Mm"], inplace=True)

# not re-run this cell buz we have drop Satellite_Rainfall_Mm column


* **Some negative values in some columns, we will replace it Nan, then handle NAN values..**

In [27]:
cleaned_data.loc[cleaned_data["Uv_Index"] < 0, "Uv_Index"] = np.nan
cleaned_data.loc[cleaned_data["Solar_Radiation_Dni"] < 0, "Solar_Radiation_Dni"] = np.nan

cleaned_data.loc[cleaned_data["Specific_Humidity"] < 0, "Specific_Humidity"] = np.nan
cleaned_data.loc[cleaned_data["Relative_Humidity"] < 0, "Relative_Humidity"] = np.nan

cleaned_data.loc[cleaned_data["Rainfall_Mm"] < 0, "Rainfall_Mm"] = np.nan
cleaned_data.loc[cleaned_data["Surface_Pressure"] < 0, "Surface_Pressure"] = np.nan

cleaned_data.loc[cleaned_data["Wind_Speed_Max_2m"]  < 0, "Wind_Speed_Max_2m"] = np.nan
cleaned_data.loc[cleaned_data["Wind_Speed_Min_2m"]  < 0, "Wind_Speed_Min_2m"] = np.nan
cleaned_data.loc[cleaned_data["Wind_Speed_Range_2m"] < 0, "Wind_Speed_Range_2m"] = np.nan
cleaned_data.loc[cleaned_data["Wind_Speed_Min_10m"] < 0, "Wind_Speed_Min_10m"] = np.nan
cleaned_data.loc[cleaned_data["Wind_Direction_Deg"] < 0, "Wind_Direction_Deg"] = np.nan


In [28]:
# cleaned_data.isna().sum()

# not re-run this if you rerun here you see all see 0 nan values

Location                     0


Year                         0

Uv_Index                 41070


Solar_Radiation_Dni      41070


Temperature_Avg              0


Dew_Point_Temp               0


Temperature_Range            0


Temperature_Max              0


Temperature_Min              0


Specific_Humidity            20


Relative_Humidity            20


Rainfall_Mm                  20


Surface_Pressure             20


Wind_Speed_Max_2m            20


Wind_Speed_Min_2m            20


Wind_Speed_Range_2m          20


Wind_Direction_Deg           20


Wind_Speed_Min_10m           20


Wet_Bulb_Temperature         0


Date                         0


Month                        0


Day                          0


dtype: int64



***130230 - 41070  = 89160***


***are the total missing values in Uv_Index, Solar_Radiation_Dni columns***

<h3 style=" 
           background:#d9ecf2; 
           color:#045270; 
           font-size:40px; 
           padding:10px;
           border-radius:12px;
           display: inline-block;">
     <b>Dealing with NAN Values</b>
</h3>

Missing values were handled using a `location-wise`, `time-based interpolation strategy`, which is a standard industry practice for ***meteorological time-series data***..

Instead of applying global statistical measures such as `mean or median`, missing observations were estimated based on adjacent temporal values within the same location. 

This approach `preserves seasonal trends`, `daily variability`, and `physical realism `while ***minimizing information loss***.

* ***Set Date as Time Index***

In [29]:
cleaned_data = cleaned_data.set_index("Date")

* ***Sort Data by Location and Date***

In [30]:
cleaned_data = cleaned_data.sort_values(["Location", "Date"])

* ***Identify Columns Requiring Imputation***

In [31]:
cols_to_interpolate = [
    
    "Uv_Index",
    "Solar_Radiation_Dni",
    "Specific_Humidity",
    "Relative_Humidity",
    "Rainfall_Mm",
    "Surface_Pressure",
    "Wind_Speed_Max_2m",
    "Wind_Speed_Min_2m",
    "Wind_Speed_Range_2m",
    "Wind_Direction_Deg",
    "Wind_Speed_Min_10m"
]


* ***Apply Location-Wise Time-Based Interpolation***

In [32]:
for col in cols_to_interpolate:
    cleaned_data[col] = (

        cleaned_data.groupby("Location")[col].transform(lambda x: x.interpolate(method="time"))
        
    )


* ***Handle Boundary Missing Values***

In [33]:
for col in cols_to_interpolate:
    cleaned_data[col] = (
        
        cleaned_data.groupby("Location")[col].transform(lambda x: x.bfill().ffill())

    )


* ***Final Missing Value Validation***

In [34]:
cleaned_data.isna().sum()

Location                0
Year                    0
Uv_Index                0
Solar_Radiation_Dni     0
Temperature_Avg         0
Dew_Point_Temp          0
Temperature_Range       0
Temperature_Max         0
Temperature_Min         0
Specific_Humidity       0
Relative_Humidity       0
Rainfall_Mm             0
Surface_Pressure        0
Wind_Speed_Max_2m       0
Wind_Speed_Min_2m       0
Wind_Speed_Range_2m     0
Wind_Direction_Deg      0
Wind_Speed_Min_10m      0
Wet_Bulb_Temperature    0
Month                   0
Day                     0
Month_Num               0
dtype: int64

* ***Restore Date as a Column***

In [35]:
cleaned_data = cleaned_data.reset_index()

In [36]:
cleaned_data.sample(5)

Date                Location  Year  Uv_Index  \
15991  1998-02-16  Hub LasbelaBalochistan  1998      1.08   
119492 1996-04-04             mirpur khas  1996      0.98   
127477 2018-02-13             mirpur khas  2018      1.21   
5278   2004-06-14                   Gharo  2004      2.44   
116590 2023-12-20                  Thatta  2023      0.86   

        Solar_Radiation_Dni  Temperature_Avg  Dew_Point_Temp  \
15991                 26.32            20.55           12.98   
119492                23.44            26.63           17.54   
127477                19.33            22.69           13.21   
5278                  18.18            34.11           23.21   
116590                14.80            19.99           -1.66   

        Temperature_Range  Temperature_Max  Temperature_Min  \
15991                5.08            23.30            18.22   
119492              12.76            33.47            20.71   
127477              12.93            29.77            16.84   
5278                10.91            40.49            29.58   
116590              15.27            28.79            13.52   

        Specific_Humidity  Relative_Humidity  Rainfall_Mm  Surface_Pressure  \
15991                9.33              62.58         0.00            100.65   
119492              12.82              63.07         0.00            100.52   
127477               9.63              61.92         0.00            101.83   
5278                18.24              57.02         1.18             99.51   
116590               3.31              24.55         0.00            101.79   

        Wind_Speed_Max_2m  Wind_Speed_Min_2m  Wind_Speed_Range_2m  \
15991                9.03               4.94                 4.09   
119492               5.50               2.16                 3.34   
127477               3.08               0.50                 2.58   
5278                 6.16               3.00                 3.16   
116590               3.54               1.08                 2.46   

        Wind_Direction_Deg  Wind_Speed_Min_10m  Wet_Bulb_Temperature  \
15991                263.0                6.40                 16.76   
119492               252.4                3.28                 22.09   
127477               292.8                0.56                 17.95   
5278                 269.7                3.75                 28.66   
116590                54.9                2.25                  9.16   

           Month        Day  Month_Num  
15991   February     Monday          2  
119492     April   Thursday          4  
127477  February    Tuesday          2  
5278        June     Monday          6  
116590  December  Wednesday         12

`This imputation strategy aligns with industry practices for meteorological time-series data and avoids artificial value generation.`

In [37]:
# Export cleaned file into csv

cleaned_data.to_csv("cleaned_weather_data.csv", index=False)



<h3 style=" 
           background:#d9ecf2; 
           color:#045270; 
           font-size:40px; 
           padding:10px;
           border-radius:12px;
           display: inline-block;">
     <b>Now the dataset is cleaned no Missing values :)</b>
</h3>
